### PR TITLE
Add testing infrastructure and test cases for request/response.

### PR DESCRIPTION
**Pull request by ericlu88**
This PR adds testing infrastructure to facilitate writing end-to-end test cases for request/response.

For test cases, the routine is to create another `node` process that host the server(`TestHelper.ts` has some helper function for this), and let the test cases act as client to hit the server. When the test server boost up, it uses `mockery` to hijack the dependency to `blpapi-wrapper.ts` and substitute it with a mocking layer `mock-blpapi-wrapper.ts`, which behaves according to an environmental parameter set by the main process. See `test-request-response.ts` as a usable example.

This is the first bit of issue #18. The next flow will be adding subscription support in the test cases.

### CODE REVIEW

File: index.ts [hunk 1]
```diff
@@ -90,6 +98,10 @@ if (conf.get('websocket.ws.enable')) {
     serverWS.listen(conf.get('websocket.ws.port'));
     serverWS.on('listening', (): void => {
         logger.info('ws server listening on', conf.get('websocket.ws.port') );
+        process.send({  // signal parent process that WS server is ready(for testing)
+            sType: 'WS',
+            message: 'server startup'
+        });
```
Comment (ericvw) [inline index.ts]: As discussed, these separate `process.send` will be funnel-down into a single `Promise.all` call to be resolved and send back a signal to the parent process.

File: lib/test-infrastructure/mock-blpapi-wrapper.ts [hunk 2]
```diff
@@ -0,0 +1,99 @@
+/// <reference path='../../typings/tsd.d.ts' />
+
+import events = require('events');
+import assert = require('assert');
```
Comment (ericvw) [inline lib/test-infrastructure/mock-blpapi-wrapper.ts]: nit: sort imports

File: lib/test-infrastructure/mock-blpapi-wrapper.ts [hunk 3]
```diff
@@ -0,0 +1,99 @@
...
+            p = p.then((): void => {
+                    assert(this[s], 'Invalid operation ' + s);
+                    this[s].call(this, cb);
+                });
+        });
+    }
+
+    // Not yet implemented
+    subscribe(subscriptions: BAPI.Subscription[], cb?: (err: any) => void): Promise<void> {
+        throw new Error();
```
Comment (ericvw) [inline lib/test-infrastructure/mock-blpapi-wrapper.ts]: Move `Not yet implemented` into the `Error()` as the message.

File: lib/test-infrastructure/mock-blpapi-wrapper.ts [hunk 4]
```diff
@@ -0,0 +1,99 @@
...
+    }
+
+    // Not yet implemented
+    unsubscribe(subscriptions: BAPI.Subscription[]): void {
+        throw new Error();
```
Comment (ericvw) [inline lib/test-infrastructure/mock-blpapi-wrapper.ts]: Move `Not yet implemented` into the `Error()` as the message.

**Comment by ericvw** [comment]
How do we feel about renaming `test-infrastructure` to just `test`?

File: lib/test-infrastructure/mock-blpapi-wrapper.ts [hunk 5]
```diff
@@ -0,0 +1,99 @@
...
+               Promise.reject(new Error('Can not stop blpsession'));
+    }
+
+    // Supported operations:
+    // sendPartialData, sendFinalData, sendError, terminateSession
+    request(uri: string, name: string, request: any, cb: BAPI.IRequestCallback): void {
+        assert(this.instructions.request, 'No request instructions.');
+        var p: Promise<void> = Promise.resolve();
+        _(this.instructions.request).forEach((s: string): void => {
+            p = p.then((): void => {
```
Comment (ericvw) [inline lib/test-infrastructure/mock-blpapi-wrapper.ts]: Would using callbacks with either `setTimeout` or `process.nextTick` be more readable here?  I understand that Promises could be used for this, but a Promise typically returns a value as a result of some async operation.  It feels a bit awkward at first glance - still mulling it over though :).

File: lib/test-infrastructure/test-helper.ts [hunk 6]
```diff
@@ -0,0 +1,47 @@
...
+    server.stdout.on('data', (data: any): void => {
+      console.log('stdout: ' + data);
+    });
+    server.stderr.on('data', (data: any): void => {
+      console.log('stderr: ' + data);
+    });
+    server.on('close', (data: any): void => {
+      console.log('child process exited with code ' + data);
+    });
+*/
```
Comment (ericvw) [inline lib/test-infrastructure/test-helper.ts]: Can this comment be removed?

File: lib/test-infrastructure/test-index.ts [hunk 7]
```diff
@@ -0,0 +1,20 @@
+/// <reference path='../../typings/tsd.d.ts' />
+
+// Setup mock blpapi-wrapper
+import assert = require('assert');
+import mockery = require('mockery');
+import MockWrapper = require('./mock-blpapi-wrapper');
+mockery.registerMock('../blpapi-wrapper', MockWrapper);
```
Comment (ericvw) [inline lib/test-infrastructure/test-index.ts]: nit: Add a line space between the imports and working code.

File: lib/test-infrastructure/mock-blpapi-wrapper.ts (hunk 5 above)
Comment (ericlu88) [inline lib/test-infrastructure/mock-blpapi-wrapper.ts]: As discussed, use `Promise.each()` for this functionality. Unfortunately, `.each()` is not exposed through the `bluebird` typing files. I've PR them for this change(borisyankov/DefinitelyTyped#3721)

File: lib/test-infrastructure/test-helper.ts (hunk 6 above)
Comment (ericlu88) [inline lib/test-infrastructure/test-helper.ts]: This is kind of useful for debugging purpose(ie, pipe the child process's stdout to the main process to figure out what's going on). Maybe add const to control whether to see it and set it if needed?

**Comment by ericlu88** [comment]
`lib/test` is kind of confusing to me, would prefer some other names.

File: lib/test-infrastructure/test-helper.ts (hunk 6 above)
Comment (ericvw) [inline lib/test-infrastructure/test-helper.ts]: Let's use the `debug` module instead to control this.

**Comment by ericvw** [comment]
Instead of `lib/test`, why don't we move the JS into the `test` folder instead.  I quick scan of other NodeJS projects do this as well.  We may need to restructure our tests to go within a subdirectory.

**Comment by ericvw** [comment]
`should` just released 5.0.1.

**Comment by ericlu88** [comment]
Updated `should` to 5.0.1

File: index.ts [hunk 8]
```diff
@@ -93,4 +106,18 @@ if (conf.get('websocket.ws.enable')) {
     });
     var wss = new webSocket.Server({ server: serverWS.server });
     wss.on('connection', webSocketHandler.wsOnConnect);
+    pServers.push(new Promise<void>((resolve: () => void,
+                                     reject: (error: Error) => void): void => {
+        serverWS.on('listening', (): void => {
+            resolve();
+        });
+    }));
```
Comment (thmiceli) [inline index.ts]: Very similar code is repeated 3 times with the only difference being the restify server variable.
I believe there is an opportunity here for refactoring into a function.

Comment (ericvw) [inline index.ts]: +1

File: test/lib/mock-blpapi-wrapper.ts [hunk 9]
```diff
@@ -0,0 +1,96 @@
...
+import Promise = require('bluebird');
+import _ = require('lodash');
+import BAPI = require('../../lib/blpapi-wrapper');
+
+export interface IInstruction {
+    start?: boolean;    // Default behavior is to start session successfully
+    stop?: boolean;     // Default behavior is to stop session successfully
+    request: string[];
+    //subscribe?: string[];
+    //unsubscribe?: string[];
```
Comment (thmiceli) [inline test/lib/mock-blpapi-wrapper.ts]: Should the two comments above be removed?

Comment (yupinghu) [inline test/lib/mock-blpapi-wrapper.ts]: I think an additional comment indicating they are not yet implemented might be more warranted. :)

**Comment by yupinghu** [comment]
As we discussed, I think ultimately the mock wrapper and the real wrapper should be peers and share some code. It doesn't need to happen for this PR, but if deferred to a later change please do create an issue to make sure we don't forget about it.

Overall I think this PR makes sense. I'm still looking over the particulars to make sure I understand it and may have more comments.

**Comment by ericlu88** [comment]
close this PR for now. Will create a new one since we decided to change path.

**Pull request closed without merging**